source/common/tracing/lightstep_tracer_impl.cc

```diff
@@ -15,103 +16,109 @@
+}
 
-void LightStepSpan::finishSpan() { span_.Finish(); }
+void LightStepDriver::LightStepTransporter::onSuccess(Http::MessagePtr&& response) {
+  try {
+    Grpc::Common::validateResponse(*response);
 
-void LightStepSpan::setOperation(const std::string& operation) {
-  span_.SetOperationName(operation);
+    Grpc::Common::chargeStat(*driver_.cluster(), lightstep::CollectorServiceFullName(),
+                             lightstep::CollectorMethodName(), true);
+    // http://www.grpc.io/docs/guides/wire.html
+    // First 5 bytes contain the message header.
+    response->body()->drain(5);
+    Buffer::ZeroCopyInputStreamImpl stream{std::move(response->body())};
+    if (!active_response_->ParseFromZeroCopyStream(&stream)) {
+      throw EnvoyException("Failed to parse LightStep collector response");
+    }
+    active_callback_->OnSuccess();
+  } catch (const Grpc::Exception& ex) {
+    Grpc::Common::chargeStat(*driver_.cluster(), lightstep::CollectorServiceFullName(),
+                             lightstep::CollectorMethodName(), false);
+    active_callback_->OnFailure(std::make_error_code(std::errc::network_down));
+  }
 }
 
-void LightStepSpan::setTag(const std::string& name, const std::string& value) {
-  span_.SetTag(name, value);
+void LightStepDriver::LightStepTransporter::onFailure(Http::AsyncClient::FailureReason) {
+  Grpc::Common::chargeStat(*driver_.cluster(), lightstep::CollectorServiceFullName(),
+                           lightstep::CollectorMethodName(), false);
+  active_callback_->OnFailure(std::make_error_code(std::errc::network_down));
 }
 
-void LightStepSpan::injectContext(Http::HeaderMap& request_headers) {
-  lightstep::BinaryCarrier ctx;
-  tracer_.Inject(context(), lightstep::CarrierFormat::LightStepBinaryCarrier,
-                 lightstep::ProtoWriter(&ctx));
-  const std::string current_span_context = ctx.SerializeAsString();
-  request_headers.insertOtSpanContext().value(
-      Base64::encode(current_span_context.c_str(), current_span_context.length()));
-}
+LightStepDriver::LightStepMetricsObserver::LightStepMetricsObserver(LightStepDriver& driver)
+    : driver_(driver) {}
 
-SpanPtr LightStepSpan::spawnChild(const Config&, const std::string& name, SystemTime start_time) {
-  lightstep::Span ls_span = tracer_.StartSpan(
-      name, {lightstep::ChildOf(span_.context()), lightstep::StartTimestamp(start_time)});
-  return SpanPtr{new LightStepSpan(ls_span, tracer_)};
+void LightStepDriver::LightStepMetricsObserver::OnSpansSent(int num_spans) {
+  driver_.tracerStats().spans_sent_.add(num_spans);
 }
 
-LightStepRecorder::LightStepRecorder(const lightstep::TracerImpl& tracer, LightStepDriver& driver,
-                                     Event::Dispatcher& dispatcher)
-    : builder_(tracer), driver_(driver) {
+LightStepDriver::TlsLightStepTracer::TlsLightStepTracer(
+    std::shared_ptr<lightstep::LightStepTracer>&& tracer, LightStepDriver& driver,
```

Comment:
nit: moving shared_ptr is a little bit of an anti-pattern IMO. I would let compiler optimize this and just pass normally.

Reply:
Changed.